mava/networks/utils/sable/encoder_decoder_fns.py

```diff
@@ -0,0 +1,271 @@
+        chunk_size=obs.shape[1],
+    )
+
+    # Apply the decoder
+    _ = autoregressive_act(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        legal_actions=legal_actions,
+        hstates=hstates[1],
+        step_count=step_count,
+        key=key,
+    )
+
+    return v_loc
+
+
+def execute_encoder_fn(
+    encoder: nn.Module,
+    obs: chex.Array,
+    decayed_hstate: chex.Array,
+    step_count: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Chunkwise encoding for Non-memory Sable and for discrete action spaces."""
+    B, C = obs.shape[:2]
+    v_loc = jnp.zeros((B, C, 1))
+    obs_rep = jnp.zeros((B, C, encoder.net_config.embed_dim))
+
+    # Apply the encoder per chunk
+    num_chunks = C // chunk_size
+    for chunk_id in range(0, num_chunks):
+        start_idx = chunk_id * chunk_size
+        end_idx = (chunk_id + 1) * chunk_size
+        chunk_obs = obs[:, start_idx:end_idx]
+        chunk_step_count = step_count[:, start_idx:end_idx]
+        chunk_v_loc, chunk_obs_rep, decayed_hstate = encoder.recurrent(
+            chunk_obs, decayed_hstate, chunk_step_count
+        )
+        v_loc = v_loc.at[:, start_idx:end_idx].set(chunk_v_loc)
+        obs_rep = obs_rep.at[:, start_idx:end_idx].set(chunk_obs_rep)
+
+    return v_loc, obs_rep, decayed_hstate
+
+
+def autoregressive_act(
+    decoder: nn.Module,
+    obs_rep: chex.Array,
+    hstates: chex.Array,
+    legal_actions: chex.Array,
+    step_count: chex.Array,
+    key: chex.PRNGKey,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    B, A, N = legal_actions.shape
+
+    shifted_actions = jnp.zeros((B, A, N + 1))
+    shifted_actions = shifted_actions.at[:, 0, 0].set(1)
+
+    output_action = jnp.zeros((B, A, 1))
+    output_action_log = jnp.zeros_like(output_action)
+
+    # Apply the decoder autoregressively
+    for i in range(A):
+        logit, hstates = decoder.recurrent(
+            action=shifted_actions[:, i : i + 1, :],
+            obs_rep=obs_rep[:, i : i + 1, :],
+            hstates=hstates,
+            step_count=step_count[:, i : i + 1],
+        )
+        masked_logits = jnp.where(
+            legal_actions[:, i : i + 1, :],
+            logit,
+            jnp.finfo(jnp.float32).min,
+        )
+        distribution = distrax.Categorical(logits=masked_logits)
+        key, sample_key = jax.random.split(key)
+        action, action_log = distribution.sample_and_log_prob(seed=sample_key)
+        output_action = output_action.at[:, i, :].set(action)
+        output_action_log = output_action_log.at[:, i, :].set(action_log)
+
+        update_shifted_action = i + 1 < A
+        shifted_actions = jax.lax.cond(
+            update_shifted_action,
+            lambda action=action, i=i, shifted_actions=shifted_actions: shifted_actions.at[
+                :, i + 1, 1:
+            ].set(jax.nn.one_hot(action[:, 0], N)),
+            lambda shifted_actions=shifted_actions: shifted_actions,
+        )
```

Comment:
Same as MAT update, much cleaner

```suggestion
        # Adds all except the last action to shifted_actions, as it is out of range
        shifted_action = shifted_action.at[:, i + 1, 1:].set(
            jax.nn.one_hot(action[:, 0], N), mode="drop"
        )
```